### PR TITLE
Add better float generator

### PR DESCRIPTION
**Pull request by ulugbekna**
- Replaced the QCheck standard float generator with a more compiler testing-oriented one, which generates corner cases more often
- Refactored integer generator: we do not to use `'a arbitrary`, as we can directly build a `Gen.t` generator. 

TODO: 
1. Pick a good ratio of corner cases to random floats. We can decide this together.
2. Implement the idea of Jan's reviewer: 

> a folklore approach to generate arbitrary integers (suggested to me by a reviewer) is to first generate an arbitrary list of numbers, and then later with, say 50% chance choose from this list, or with 50% chance generate a "fresh" random number. Such a generator will have a reasonable chance of outputting duplicate numbers (a common boundary case).

### CODE REVIEW

File: src/efftester.ml [hunk 1]
```diff
@@ -173,9 +173,11 @@ let efftostr ((ef, ev) : eff) = Printf.sprintf "(%B,%B)" ef ev
 let toOCaml ?(typeannot = true) term =
   let rec littoOcamlSB sb = function
     | LitUnit -> Printf.bprintf sb "()"
-    | LitInt i -> if i < 0 then Printf.bprintf sb "(%d)" i else Printf.bprintf sb "%d" i
+    | LitInt i -> if i <= 0 then Printf.bprintf sb "(%d)" i else Printf.bprintf sb "%d" i
```
Comment (gasche) [inline src/efftester.ml]: What's the reason for this particular change?

**Review by gasche — COMMENTED** [review]
This looks fine to me -- see minor question on the int literal printing.

File: src/efftester.ml (hunk 1 above)
Comment (ulugbekna) [reply src/efftester.ml]: The reason is the same as for the float change. I am just being precautious if we at some point want to add test for `-0`. I can see, however, that it might clutter the output, and we can remove it at least for now.

Comment (ulugbekna) [reply src/efftester.ml]: The bug caused by `-0.` without parentheses was quite perplexing for me even though it turned out to be trivial.

**Comment by gasche** [comment]
But there is no `-0` integer, while there is a "negative zero" float value; so in the integer case, you are adding parentheses around `0`.

File: src/efftester.ml (hunk 1 above)
Comment (jmid) [reply src/efftester.ml]: For floats I believe `-0.` is a valid concern, but for integer literals I don't follow you.

**Comment by gasche** [comment]
In the float case, ideally we would parenthesize `-0.` but not `0.`; this can be done, but it is actually quite delicate to implement, given that both numbers will claim to be equal (the comparison operators do not distinguish them, following the spec.); a way to know if a floating-point `x` is negative is to check `copysign 1. x = -1. ` (`copysign a b` will return a number with the absolute value of `a` and the sign of `b`). I think it's simpler to just have parentheses around all negative numbers *and* `0.`, as your implementation does.

**Comment by ulugbekna** [comment]
Regarding TODO no 2, I created a generator that uniformly picks from two possible generators:
1. Ordinary `float_gen` from earlier commit
2. A generator that takes values from cache (a list of values that were previously generated by `float_gen`)
After picking a float from one of the two generators, the generated value is put in cache if it's not already there. 

I have created like this in `efftester.ml`:

```ocaml
let float_cache = ref []
let clear_float_cache () = float_cache := []

let float_gen_with_rep : float Gen.t =
  let open Gen in
  let from_cache_gen rs = (oneofl !float_cache) rs in
  fun rs ->
    match !float_cache with
    | [] ->
      let fl = float_gen rs in
      float_cache := fl :: !float_cache;
      fl
    | _xs ->
      let fl = (oneofl [ float_gen; from_cache_gen ] |> join) rs in
      if List.mem fl !float_cache
      then fl
      else (
        float_cache := fl :: !float_cache;
        fl)
;;
```
But this solution seems unidiomatic to me because of "abuse" transparency of type `'a Gen.t`. However, I have already spent quite a lot of time trying to make it idiomatic that I am not sure if I should leave it as it is. 

Also, we could put the cache in our generator's closure, assuming that we do not rerun the tests.

**Comment by ulugbekna** [comment]
Also, would we want a similar cache-based generator for ints?

**Comment by ulugbekna** [comment]
Also, I changed the initial idea, which, in my understanding, was "generate a list of random values first, then uniformly either pick a value from that list or generate a new one". 

My idea is to generate values, cache them, and each further generation uniformly either picks from that cache or generate a new value. 

I was motivated to change the idea by the fact that we do not know the size of the generated program. In, say, a small generated program, why to  generate an arbitrary-length list of values from which only a few values end up being used. That would increase the testing time and decrease the probability of reuse of generated values.

**Comment by ulugbekna** [comment]
If you want to test `float_gen_with_rep`, this might be useful: 
```ocaml
let float_test_run () =
  let arb_float =
    make ~print:string_of_float ~collect:string_of_float float_gen_with_rep
  in
  let float_test =
    Test.make ~count:100 arb_float (fun f ->
        print_endline @@ "float_test_run: " ^ string_of_float f;
        true)
  in
  QCheck_runner.run_tests [ float_test ]
;;
```

**Comment by gasche** [comment]
It took me a bit of thinking to convince my that caching using hidden mutable state was actually okay. Here is what I find convincing:

- the changes to the mutable state are only driven by random-generator draws, nothing else (in particular, running a test from a fix seed several times, we will always observe the same changes each time, so we remain deterministic (for a fixed seed))

- the changes to the mutable state are only observed by calling the float random generator, they cannot affect any other part of the program

Accordingly, I think that it would be important for the mutable state to live *within* the generator definition, so that we know by scoping that it cannot be accessed from the outside. (This requires giving up on the ability to "clear" the generator, which I think could endanger the deterministic behavior I mentioned above.) This would be similar to this "fresh integer generator":

```ocaml
(* [fresh_id ()] returns a new integer each time it is called, starting from 0 *)
let fresh_id =
  let counter = ref (-1) in
  fun () ->
    incr counter;
    !counter
```

**Comment by gasche** [comment]
Regarding the int/float question, I still think that the right solution would be to test `if i < 0` in the integer case, and `if x <= 0.` in the floating-point case (with a comment to indicate that the inequality is large because `-0.` needs to be parenthesized). Should we discuss that point more?

**Comment by jmid** [comment]
> ```ocaml
> ...
>     | _xs ->
>       let fl = (oneofl [ float_gen; from_cache_gen ] |> join) rs in

This part reads a bit strange to me. `Gen.oneofl` is a combinator for choosing among a list of values. There's another QCheck combinator `Gen.oneof` for choosing among a list of generators.
I think it could be more clearly stated using that instead.

I have been teaching a course on QuickCheck with OCaml and QCheck at SDU.
The material from last year is available here: http://janmidtgaard.dk/quickcheck/
In particular, I cover the various combinators in lecture 3:
http://janmidtgaard.dk/quickcheck/slides/lec03.pdf  (p.42 and onwards in pdf, labeled 21/47 in the bottom right corner). Since you have your hands in QCheck combinators already, you might appreciate it 🙂

**Comment by jmid** [comment]
> Accordingly, I think that it would be important for the mutable state to live _within_ the generator definition, so that we know by scoping that it cannot be accessed from the outside. (This requires giving up on the ability to "clear" the generator, which I think could endanger the deterministic behavior I mentioned above.) This would be similar to this "fresh integer generator":
> ...

I agree with Gabriel that keeping the state local is important to avoid later surprises.
I think if you declare both the generator and the "clear" simultaneously, one can have both private mutable state and the ability to clear, like in the fresh (type)var generators
https://github.com/jmid/efftester/blob/5bef29e3c9c7f37a3d6962083b13f2c7775384e8/src/efftester.ml#L53

**Comment by ulugbekna** [comment]
Thanks for the comments! I have went over them a couple of times. This is what I came up with as a result: 

```ocaml
(* a function that gives us operations over its local value, which is inaccessible globally *)
let new_cache () =
  let cache = ref [] in
  let to_cache fl = cache := fl :: !cache in
  let get_cache_lst () = !cache in
  let clear_cache () = cache := [] in
  (to_cache, get_cache_lst, clear_cache)
;;
``` 
Below, there are `float_gen_with_rep` generator, which keeps the same cache for different runs of test, and `float_gen_with_rep_thunk`, which on call will return a generator with new cache. The latter allows us to have a deterministic generator with  a "fresh-cache" :-) The last function is a test that demonstrates that the latter function: 
- is deterministic (generates same values for same seed)
- has clean cache each time

There is also a counter 

```ocaml
let rec print_list lst =
  print_string "[ ";
  List.iter
    (fun fl ->
      print_float fl;
      print_string "; ")
    lst;
  print_string "]\n"
;;

(* Generate a possibly repeated floating-point number *)
let float_gen_with_rep : float Gen.t =
  let open Gen in
  let to_cache, get_cache_lst, _clear_cache = new_cache () in
  let from_cache_gen rs = (oneofl @@ get_cache_lst ()) rs in
  fun rs ->
    print_list (get_cache_lst ());
    match get_cache_lst () with
    | [] ->
      let fl = float_gen rs in
      to_cache fl;
      fl
    | _xs ->
      let fl = (oneof [ float_gen; from_cache_gen ]) rs in
      if List.mem fl (get_cache_lst ())
      then fl
      else (
        to_cache fl;
        fl)
;;

let float_gen_with_rep_thunk () =
  let open Gen in
  let to_cache, get_cache_lst, _clear_cache = new_cache () in
  let from_cache_gen rs = (oneofl @@ get_cache_lst ()) rs in
  fun rs ->
    print_list (get_cache_lst ());
    match get_cache_lst () with
    | [] ->
      let fl = float_gen rs in
      to_cache fl;
      fl
    | _xs ->
      let fl = (oneof [ float_gen; from_cache_gen ]) rs in
      if List.mem fl (get_cache_lst ())
      then fl
      else (
        to_cache fl;
        fl)
;;

let float_test_run () =
  let arb_float =
    make ~print:string_of_float ~collect:string_of_float (float_gen_with_rep_thunk ())
  in
  let float_test =
    Test.make ~count:100 arb_float (fun f ->
        print_endline @@ "float_test_run: " ^ string_of_float f;
        true)
  in
  QCheck_runner.run_tests ~rand:(Random.State.make [| 1 |]) [ float_test ]
;;
```

Let me know if it's easier for you, if I actually make commits instead of putting code in the comments.

**Comment by jmid** [comment]
Nice work! I think I like the `float_gen_with_rep_thunk`-version best,
as the thunk argument signals "side effect" (cache allocation) in the type.
If one wishes to use floats generators with different caches (not sure whether
this would make sense in efftester, but still) this is possible 🙂

**Comment by ulugbekna** [comment]
Trying to use `float_gen_with_rep_thunk` in our `dep_term_gen`, I came to a conclusion:

The more local a side effect is, the harder it is to propagate it to upper layers. If the cache is local to `float_gen_with_rep_thunk`, all functions that use the generator created by the thunk need to have have the _same_ generator with the _same_ cache. This, in my understanding, means that we will need thunks of all functions that use float_generator explicitly or implicitly (have a function that uses the generator). 

This is our slightly improved float generator thunk: 
```ocaml
let float_gen_with_rep_thunk () =
  let to_cache, get_cache_lst, _ = new_cache () in
  let from_cache_gen rs = (Gen.oneofl @@ get_cache_lst ()) rs in
  fun rs ->
    match get_cache_lst () with
    | [] ->
      Gen.map
        (fun fl ->
          to_cache fl;
          fl)
        float_gen
        rs
    | _xs ->
      Gen.map
        (fun fl ->
          if List.mem fl (get_cache_lst ())
          then fl
          else (
            to_cache fl;
            fl))
        (Gen.oneof [ float_gen; from_cache_gen ])
        rs
```

There is a function `literalGen` that calls our thunk and gets the generator. This means that on every call to `literalGen`, we will get a new cache. We can make `literalGen` a thunk too, initializing a single `float_gen_with_rep` in its closure: 

```ocaml
let literal_gen_thunk () =
  let float_gen_with_rep = float_gen_with_rep_thunk () in
  let rec literalGen t eff size =
    match t with
    | Unit -> Gen.return LitUnit
    | Int -> Gen.map (fun i -> LitInt i) int_gen
    | Float -> Gen.map (fun f -> LitFloat f) float_gen_with_rep
    | Bool -> Gen.map (fun b -> LitBool b) Gen.bool
    | String -> Gen.map (fun s -> LitStr s) stringGen
    | List (Typevar _) -> Gen.return (LitList [])
    | List t ->
      if size = 0
      then Gen.return (LitList [])
      else
        Gen.map
          (fun ls -> LitList ls)
          (Gen.list_size (Gen.int_bound (sqrt size)) (literalGen t eff (sqrt size)))
    (*     (Gen.list_size (Gen.int_bound (size/2)) (literalGen t eff (size/2))) *)
    (* FIXME: - one element should/can have effect, if 'eff' allows *)
    (*        - list items should be able to contain arbitrary effectful exps *)
    | Typevar _ -> failwith "literalGen: typevar arg. should not happen"
    | Fun _ -> failwith "literalGen: funtype arg. should not happen"
  in
  literalGen
;;
```

Nevertheless, we face the same problem as with the float gen thunk: there is a function `litRules` that calls the literalGen thunk and is called on every generation. Propagation of a too local effect can be costly (?)

I am thinking if we should stick with simple `float_gen_with_rep` and simply clear the cache after every full test run, propagate the local effect further (complex code or better solution that I do not see yet), or try implementing the initial idea 😅.

**Comment by ulugbekna** [comment]
Anyway, I will keep experimenting and see what we can do. Have a nice weekend!

**Comment by jmid** [comment]
> Anyway, I will keep experimenting and see what we can do. 

I see your point. Couldn't one "just" let-bind a top-level thunk application:
```
  let my_cached_float_gen = float_gen_with_rep_thunk ()
```
We'll also need a top-level `clear_local_float_cache` to ensure that repeated runs
deterministically produces the same test inputs.

> Have a nice weekend!

You too! 🙂

**Comment by gasche** [comment]
I see three options:

1. Have a toplevel reset that we are careful to call after each toplevel test. This is the simplest option to implement, but I'm not fond of the fact that it makes it easy to shoot ourselves in the foot by forgetting to clear the cache. (Basically we have all the problems of global mutable state.)

2. Thread an `env` or `context` parameter throughout the generators, that carries the state. (Each test would start by generating its own state value.) Our generators would thus be of the type `env -> 'a Gen.t`. 

   One annoyance with that is that `QCheck` does not provide good support for parametrized generators, in particular it does not provide a parametrized fixpoint combinator (compare [QCheck's `fix`](https://github.com/c-cube/qcheck/blob/b14291a/src/core/QCheck.mli#L316) with the one in [random-generator](https://github.com/gasche/random-generator/blob/63dedf8/random_generator.mli#L156) for example). We could contribute a `pfix` upstream. In any case, in this specific case (the context we want to pass is a global mutable state), we don't actually need a parametrized fixpoint, because the context value never changes (it points to the cache which gets mutated; in Haskell terms, it's an Env monad to mutable shared state, rather than a State monad). But we may want to pass more things in the context in the future.

   A third annoyance is that it means changing all the generator code to pass this extra `ctx` or `env` parameter all over the place, which is invasive. This is what (3) below can avoid.

3. We could store all the generators in a common functor that is parametrized over a state value. Each test would start by instantiating this functor with a new state. This allows to do roughly the same as option (2) (with the same constraint that the generators cannot replace a state value with another, they can just mutate a single state value), but is much less invasive. The codebase would look like this:

   ```ocaml
   module type Context = sig
     val float_gen : float Gen.t
   end
   module Generators (Ctx : Context) = struct
     (* all the generators here, calling Ctx.float_gen when necessary *)
   end
   
   let test_foo () =
     let module Gens = Generators(struct let float_gen = float_gen_thunk () end) in
     let open Gens in
     ...
   ```

In my experience as an OCaml programmer, I would say that I have mixed feelings about functors. On paper they often look like a good approach to a problem, but in practice they give code that feels a bit heavy, often a bit more complex than we would like. Option (2) is much more work now (and it sounds annoying to add a new parameter all over the place) but it may result in code that is slightly more pleasant to use in the end. I would still recommend going for (3) for now (it's less work), and only switching to (2) later *if* you start getting irritated by the verbosity of the module/functor manipulation.

**Comment by ulugbekna** [comment]
I also like @gasche's third solution and will go with it for now.

**Comment by ulugbekna** [comment]
If we put generator functions into a functor `GeneratorsF`, which takes a module `Context`
as a parameter, there are other functions (shrinkers) outside the functor's scope that
need generator functions. Those shrinkers, in-turn, are needed by `'a arbitrary` creator
functions. Thus, we need to think about how shrinkers can access generators and arbitrary
generators can access shrinkers. AFAICT, there are three ways we can go from here:

1. Create an overarching module `Tester` that would enclose, afaiu now, a functor for
   generating generators module, instance of which will be passed to shrinkers functor,
   instance of which is passed to arbitraries's module. All modules and their instances
   (of generators and shrinkers) will be stored in the functor `Tester`. I think this
   level of coupling is not a very good idea.

2. Create a functor `Tester` that would take in context, which will be available to all
   modules: generators, shrinkers, functors. As the latter three modules will not be
   functors, I think we will not need to instantiate and store them. Seems like a better
   idea than the first one even though it has quite tight coupling - submodules will not
   be reusable outside `Tester`. 

3. Go with Jan's idea (again, taking a precaution to clear cache before each test).

Sorry for coming up with new problems all the time. I just wanted to discuss this before making quite large architectural changes.

**Comment by gasche** [comment]
We can have shrinkers in their own functor parametrized over the same `Context` module (a float generator), that internally call the generator-functor to get generators. The way our functors use their context, calling the same functor over the same context module several times gives functions that are inter-operable -- this is a good property.

Currently the `arbitrary` definitions are given from the generators alone, and then they are redefined using `set_shrink` to get the right shrinker values. @jmid, should we keep this design? (I'm not sure what the purpose is, maybe to avoid the shrinkers when we do not fully trust them?) It would be a bit simpler to have a single construction of arbitrary instances from (an instance of) the generators and (an instance of) the shrinkers.

I think that the tests themselves should not be in a functor; rather, each test should create its own local context and instantiate the relevant functors locally. This ensures that, from the shape of the code, we know that the mutable state of a test never pollute another test.

Roughly here is what I would have in mind:

```ocaml
module type Context = sig
  val float_gen : float gen
end

(* here the signature constraint  `: Context` guarantees
    that potential bits of toplevel mutable state maintained
    within the module are not visible outside *)
module FreshContext () : Context = struct
  ...
end

module Generators (Ctx : Context) = struct
  ... (* including a rebinding of `float_gen` for consistency *)
end

module Shrinkers (Ctx : Context) = struct
  module Gener = Generators(Ctx)
  ...
end

module Arbitrary (Ctx : Context) = struct
  module Gener = Generators(Ctx)
  module Shrink = Shrinkers(Ctx)
  ...
end

let tcheck_test () =
  let module Ctx = FreshContext() in
  let module Arb = Arbitrary(Ctx) in
  Test.make ... Arb.int_term ...
```

Ideally, this should also let you get rid of the `resetvar()` and `resettypevar()` calls in Effmain -- and in general -- if we can also move the relevant pieces of mutable state in the context.

**Comment by jmid** [comment]
> Currently the `arbitrary` definitions are given from the generators alone, and then they are redefined using `set_shrink` to get the right shrinker values. @jmid, should we keep this design? (I'm not sure what the purpose is, maybe to avoid the shrinkers when we do not fully trust them?) It would be a bit simpler to have a single construction of arbitrary instances from (an instance of) the generators and (an instance of) the shrinkers.

I think the current `set_shrink` is just a left over from experimentation.
I would go for `make ~shrink:...` to build a (full) `arbitrary` generator out of a (pure)
`gen` generator - and simultaneously set the printer and shrinker.

**Comment by ulugbekna** [comment]
I made one large commit instead of breaking it into smaller ones because I thought it is better to have commits non-breaking, i.e., smaller commits would have code that does not compile.
